### ISSUE TITLE
CogSec egress boundary: propagate confidentiality and require explicit declassification

### ISSUE THREAD

**Issue by Tristan-Stoltz-ERC** [Luminous-Dynamics/symthaea#214]
## Purpose

Make CogSec bidirectional: protect not only what remote information may change, but what cognitively derived state may leave the local confidentiality domain.

Inbound security primarily protects control integrity. Outbound security must protect confidentiality and disclosure authority.

### `MindOutput` has no security/confidentiality metadata

Current `MindOutput` contains:

- output type;
- string content;
- HDC embedding;
- confidence;
- emotional tone.

It carries no provenance, confidentiality, taint or declassification status.

### Output confidence currently reflects consciousness integration

The normal output path sets `MindOutput.confidence` from `consciousness_level`.

That value may be useful cognitive telemetry, but it is neither factual/evidential confidence nor disclosure authority.

### Social egress exports active cognitive state

The social loop periodically places `current_thought` into outbound `SocialMessage` behavior/context fields and may include neuromodulator bath state.

### Mesh egress exports active cognitive state

Mesh auto-emission derives a wisdom vector from `current_thought` and broadcasts it according to mesh policy/urgency.

Federated gradient export can also disclose local model/training information.

These are different confidentiality sinks and must not inherit “safe to send” merely because the corresponding subsystem is enabled.

## Constitutional rule

> Information cannot become less confidential merely by passing through cognition.

And:

> Generation is not declassification.

And:

> A model's choice to mention, summarize, transform, compress or embed private information does not grant permission to disclose it.

## Conservative derivation rule

A generated/derived artifact should conservatively inherit confidentiality from all security-relevant context that materially contributed to it unless dependency minimization or another qualified mechanism proves a narrower dependency set.

Conceptually:

`Public prompt + Private memory -> PrivateDerived output candidate`

not:

`Public prompt + Private memory -> Public because the answer is newly generated`.

For HDC bundles/summaries/embeddings, transformation does not remove confidentiality by itself.

### `EgressCandidate<T>`

Wrap generated/disclosure-bound data with:

- payload or opaque payload commitment;
- `CognitiveSecurityLabel`/confidentiality label;
- provenance/dependency commitments;
- destination/sink class;
- intended recipients/domain;
- purpose/reason if policy uses purpose binding;
- generation/derivation lineage;
- local sequence;
- no declassification authority embedded in payload.

### Egress sink classes

At minimum distinguish:

- local UI/display;
- private local logs;
- operator/user response channel;
- remote LLM/API provider;
- trusted organizational peer;
- social peer/network;
- public mesh/broadcast;
- federated-learning export;
- tool/service invocation payload;
- file/persistent export.

One sink's permission does not imply another's.

### `DisclosurePolicy`

Local policy determines whether a candidate label can flow to a sink.

Policy input includes:

- source confidentiality;
- destination confidentiality domain;
- recipient identity/authentication where applicable;
- purpose/context;
- consent/capability facts;
- policy/revocation epochs;
- optional data-minimization transforms.

### `Declassify` mutation

Any deliberate confidentiality reduction is a privileged CogSec transition.

A declassification request binds:

- exact artifact/effect commitment;
- current confidentiality label;
- target less-restrictive label/domain;
- exact destination/recipient class where feasible;
- reason/policy basis;
- policy root/epoch;
- authorization/revocation epochs;
- expiry/scope where applicable.

Declassification authority is not carried forward automatically to later derived artifacts unless policy explicitly says so.

## Separate confidentiality from epistemic/control integrity

A datum may be:

- highly confidential and factually uncertain;
- public and authenticated;
- private and untrusted;
- sensitive but safe to analyze locally.

Do not collapse confidentiality into trust, factual confidence, salience or authority.

### E0 — output observation

Instrument current outputs with a shadow-derived label and dependency/provenance summary without changing legacy response behavior.

### E1 — typed output envelope

Introduce a security-aware wrapper around `MindOutput` or a separate `SecuredMindOutput` at the boundary. Avoid changing the cognitive core's user-facing API in the same step if that makes non-interference difficult to prove.

### E2 — sink classification

Every output/disclosure adapter declares an explicit sink class rather than accepting a naked string/HDC value.

### E3 — shadow egress decision

Evaluate disclosure policy, record would-allow/would-deny/declassify-needed, but continue legacy egress. Count legacy disclosures after non-Allow.

### E4 — enforcement

Gate actual disclosure only after sink coverage and non-interference evidence are complete.

## Social and mesh egress

Do not assume social/mesh data are public merely because they are encoded HDC vectors.

Embeddings may preserve sensitive semantic information and should inherit conservative confidentiality from their dependencies.

For `current_thought`/wisdom emission:

- derive an egress candidate;
- bind the destination/mesh domain;
- apply confidentiality policy;
- minimize/transform only through an explicit qualified transform;
- require declassification if policy requires a confidentiality reduction.

A coarse/quantized/compressed vector is not automatically declassified.

## LLM/API boundary

Prompts sent to remote language-model providers are outbound disclosures.

Before sending context externally:

- classify provider/sink;
- propagate confidentiality from prompt + retrieved context + memory;
- minimize context where possible;
- require explicit allowed flow/declassification where needed;
- record enough provenance to audit which restricted sources were made visible.

A local backend and a remote API are distinct sink classes even if both implement the same inference trait.

## Federated export

Gradient/model-update export can leak training data/model information despite differential privacy controls.

Treat export as a confidentiality/egress decision separate from `LearningPromotion`.

DP qualification may reduce disclosure risk but does not itself authorize network release.

Bind export policy to:

- exact model/training resource;
- privacy mechanism/profile;
- privacy budget/accounting state;
- recipient/network class;
- source confidentiality constraints.

## Dependency minimization

Conservative “join all visible context” labeling is safe but may over-taint outputs.

Later higher-assurance optimization may support dependency narrowing using:

- explicit retrieval dependency sets;
- structured generation plans;
- information-flow tracking;
- verified redaction/transformation;
- isolated/public-only generation contexts.

But the rule is:

> Unproven independence cannot justify declassification.

## Declassification transforms

A transform such as redaction, aggregation or anonymization can be considered only if its security property is separately qualified.

Do not treat:

- summarization;
- paraphrasing;
- embedding;
- HDC projection;
- compression;
- encryption for another recipient

as automatic declassification.

Encryption changes transport confidentiality, not the underlying classification of the plaintext/information.

## Audit/privacy balance

Egress evidence should record commitments/labels/sinks without unnecessarily logging raw sensitive output.

Default security receipts should include:

- artifact commitment;
- source/dependency roots;
- source and target confidentiality classes;
- sink/recipient class;
- policy/declassification result;
- epochs;
- receipt outcome.

Avoid placing raw secrets into the security ledger merely to prove they were protected.

## Test obligations

1. public input + restricted memory yields at least restricted/private-derived candidate without qualified dependency narrowing;
2. summarization cannot lower confidentiality by itself;
3. HDC projection/compression cannot lower confidentiality by itself;
4. a candidate blocked for public mesh may still be allowed for local UI according to policy;
5. permission to send to one authenticated peer cannot be reused for public broadcast;
6. declassification permit is bound to exact artifact/destination and cannot release altered content;
7. revocation/policy epoch change invalidates outstanding declassification authority;
8. generated text cannot self-label as public to override trusted context;
9. remote LLM call is rejected/flagged when private context lacks an allowed provider flow;
10. local LLM inference can remain allowed under a stricter confidentiality policy than remote inference;
11. federated gradient export requires separate disclosure policy even when DP is enabled;
12. raw sensitive content does not appear in default egress receipts;
13. legacy shadow mode records disclosures after hypothetical non-Allow without changing behavior;
14. every network/file/tool disclosure sink is present in the egress census before enforcement.

## Metrics

Track exact counters such as:

- egress candidates by sink;
- confidentiality joins performed;
- would-allow/would-deny;
- declassification required/granted/denied;
- legacy disclosures after non-Allow in shadow;
- disclosures without egress decision in enforcement mode;
- missing-label egress attempts;
- remote-provider private-context attempts;
- privacy-budget exhaustion for federated export.

Critical enforcement targets:

- `egress_without_security_label == 0`
- `restricted_disclosures_without_declassification == 0`
- `disclosures_without_egress_decision == 0`

## Non-claims

The goal is not to suppress useful communication or make the system unable to discuss private information with the user who owns it. The goal is explicit **destination-aware disclosure authority** so confidentiality cannot silently disappear during reasoning, generation, projection or network transport.

**Comment by benrrr56-wq** [comment]
Making CogSec bidirectional so it also controls what cognitively derived state leaves the local confidentiality domain is the right gap to close, because inbound-only protection leaves the exfil side implicit. The declassification requirement is the load-bearing piece: it's only meaningful if 'declassified' is an explicit operation with a recorded decision, not if a tool can implicitly send derived state out because nothing blocks it. Does the boundary evaluate the destination, the sensitivity label of the derived state, or both, before anything egresses? If confidence is derived from a document and the tool pastes it to an external service, the control needs to see that derived content is confidentially-tagged and stop it, which means tracking sensitivity through transformation rather than only at ingestion. I build control/observation tooling, and the explicit-declassification-with-a-recorded-decision part is where I'd put the effort: control what an agent can do, see what actually ran, prove it with recomputable evidence. https://agentkey.us

**Comment by Tristan-Stoltz-ERC** [comment]
Cross-project composition update: Mycelix #2746 / draft PR #2747 now freezes OPSEC-000 as a **shared semantic/evidence waist**, explicitly preserving CogSec #214 as the owner of cognitive confidentiality propagation/declassification semantics rather than duplicating it.

The intended split is:

```text
Symthaea CogSec
  produces/propagates confidentiality + lineage candidates
  enforces 'generation != declassification'
        ↓ typed external references
Mycelix OPSEC waist
  binds exact subject + sink + purpose + policy/currentness evidence
        ↓
separate live disclosure/execution authority
```

Mycelix #2748 makes web acquisition the first concrete consumer and adopts the same rule for search/API egress: a public/SSRF-safe destination does not authorize a private-derived query/body/header, and remote search/LLM prompts are destination-specific disclosure sinks.

No change to #214's authority is intended: Symthaea still cannot self-declassify, and Mycelix OPSEC decision evidence is not a DispatchPermit.

**Comment by Tristan-Stoltz-ERC** [comment]
The CogSec egress rule is now explicitly reused by the OSINT hardening line rather than duplicated: Mycelix #2862 records sensitive-data lineage through generated queries/prompts/summaries/embeddings and #2863 makes declassification a qualified exact-artifact transition; Symthaea #5523/#5524 consume those semantics. Generation/embedding/compression therefore remains non-declassifying by default.
